### PR TITLE
#162949078 Users should be able to comment on articles

### PR DESCRIPTION
**Pull request by sharkdevs**
### What does this PR do?
Allows a user to comment on an article.

### Description of Task to be completed?
- [x] Setup a comments app
- [x] Add tests
- [x] Create comments API endpoint
- [x] Add Update comments API endpoint
- [x] Get all comments API
- [x] Get one comment API
- [x] DELETE Comment API
- [x] TEST Comment API
### How should this be manually tested?
1. Clone this repository on your local machine using 

    ```$```  ```git clone https://github.com/andela/ah-robotics.git```
2. Change your working directory to the project's directory using 
     
     ```$``` ```cd ah-robotics```
3. Change to the specific branch `ft-comments-162949078` using 

     ```$``` ```git checkout ft-comments-162949078```
4. Create a virtual environment using the following command: 
     ```$```  ```python3 -m virtualenv env```
5. Activate the virtual environment using 
```$``` ```source env/bin/activate```
6. Install all the requirements using 
     ```$``` ```pip install -r requirements.txt```

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
       ```$``` ```pg_ctl -D /usr/local/var/postgres start```
    
    2. Run psql interactive terminal
    
        ```$``` ```psql postgres```
    
    3. Create a new user:
    
        ```$``` ```CREATE USER sample_username WITH PASSWORD 'sample_password';```
    
    4. Grant privileges to the user:
    
        ```$``` ```ALTER USER sample_username CREATEDB;```
    
    5. Create a database:
    
       ```$``` ```CREATE DATABASE sample_database_name WITH OWNER sample_username;```
    
    6. Exit psql interactive with the following command: ```$``` ```\q```
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get cloudinary values, create an account with cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      ```$``` ```source .env```
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      ```$``` ```python manage.py makemigrations```
      
      (ii) Apply the migrations using: 
      
      ```$``` ```python manage.py migrate```

10. Run the following command to collect static files.

   ```$``` ```python manage.py collectstatic```
      
#### Testing the features

1. Run the server from the terminal using: 
    
  ```$```  ```python manage.py runserver```
2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL `http://127.0.0.1:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and then add a payload to the body using the following format:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

**Example** 
sample_username---> honululu, 
sample_email-->hanku@example.com, 
password-->test1234

5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Log in the user using the following endpoint `http://127.0.0.1:8000/api/v1/users/login/` and insert the following payload to the body section.
    
    ``` {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

**NB** Use details you created in preceding step ```4```
7. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
8. After logging in the screen should look as follows:
    
    
screenshot 2019-01-29 at 15 59 01

9. Copy the token and click on the **Headers** section. In the **Key** column, type in the word **Authorization**. And in its corresponding value in the **Value** column, write the word __token__, add a space and paste in the token that you copied from the login response. The **Headers** section should look as follows after doing so:
    
    image

10. To Create an article, insert the following URL: http://127.0.0.1:8000/api/v1/articles/. In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, post an article by using the following payload:
    
    ```
    {
        "article": {
            "title": "Sample Article",
            "description": "Sample Article Description",
            "body": "The sample article body"
        }
    }
    ```
    Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.

11. To comment on an article, use the URL http://127.0.0.1:8000/api/v1/articles/slug/comments/ where `slug` is a unique identifier for the given article. Change the method to a **POST** and add the following to the body before you hit **Send**.
    
    ```
    {
       "comment": {
         "body": "His name was my name too."
       }
    }
    ```
12. To get all comments, use the URL http://127.0.0.1:8000/api/v1/articles/slug/comments/. Change the method to a **GET** before you hitting **Send**.
13. To get a specific comment to an article, use the  URL http://127.0.0.1:8000/api/v1/articles/slug/comments/id/ where `id` is the numeric id of the comment.
14. To update a comment, use the  URL http://127.0.0.1:8000/api/v1/articles/slug/comments/id/ as explained above. However, add the update of your comment in the following format.
    
    ```
    {
       "comment": {
         "body": "My commit update."
       }
    }
    ```
15. To delete a given comment, use the  URL http://127.0.0.1:8000/api/v1/articles/slug/comments/id/ where `id` is the numeric id of the comment. Click send to delete.
    
#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#162949078](https://www.pivotaltracker.com/n/projects/2235171/stories/162949078)

### Screenshots
create a comment
screen shot 2019-02-01 at 00 08 53

get all comments:
screen shot 2019-02-01 at 00 09 45

get one comment:
screen shot 2019-02-01 at 00 10 40

update a comment
screen shot 2019-02-01 at 00 11 20

delete comment
screen shot 2019-02-01 at 00 11 46

## Questions:
Questions and comments are welcome.

### CODE REVIEW

**Comment by jomasim** [comment]
consider naming this**(feat(comments): create comments API)** commit header to 
**feat(comments): create comments api view**.  It should not be called an API it is a view in inside the API

**Comment by jomasim** [comment]
The commit messages need to be amended as well.
**- add api to create comment**
**- add api to get all comments**
consider renaming them to something like this : 
**add create comment api endpoint** and not just api

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @sharkdevs ,
I have tested this feature on my machine and it works as expected. Good work. Just fix the merge conflicts

**Comment by sharkdevs** [comment]
> Hey @sharkdevs ,
> I have tested this feature on my machine and it works as expected. Good work. Just fix the merge conflicts

Thank you @Paulstar200, working on them

**Review by waracci — COMMENTED** [review]
Hey Meshack it works as expected. Good job.

**Comment by Teatoller** [comment]
Hello @sharkdevs ,

I have been able to follow through and the instructions and realized the desired outputs.
screenshot 2019-02-04 at 10 16 30

Regards
Steven Ennis

File: authors/apps/comments/serializers.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+from rest_framework import serializers
+
+from authors.apps.profiles.serializers import ProfileSerializer
+from authors.apps.profiles.models import UserProfile
+from .models import Comment
+
+
+class Commentserializer(serializers.ModelSerializer):
```
Comment (johnwayodi) [inline authors/apps/comments/serializers.py:8]: Ensure to follow Python Naming Convention when giving names to your classes, consider renaming this class.

File: authors/apps/comments/tests/test_comments.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        self.create_article(self.sample_article)
+        res = self.create_comment(
+            self.sample_comment, self.sample_article)
+        url = self.comment_url("my-article")+'2/'
+        response = self.client.delete(url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data['comment']
+                         ['message'], "Comment deleted successfully")
+
+    def test_get_inexistent_comment(self):
```
Comment (johnwayodi) [inline authors/apps/comments/tests/test_comments.py:120]: There seems to be some grammatical error here, kindly correct the function name.

**Review by johnwayodi — COMMENTED** [review]
Hello @sharkdevs, I've noticed some issues and left some comments, kindly look into them and correct them.

**Pull request merged**